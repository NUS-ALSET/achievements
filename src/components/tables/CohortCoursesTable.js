/**
 * @file CohortCoursesTable container module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 22.02.18
 */

import React from "react";
import PropTypes from "prop-types";

import { Link, withRouter } from "react-router-dom";

import Button from "@material-ui/core/Button";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import TableSortLabel from "@material-ui/core/TableSortLabel";

import withStyles from "@material-ui/core/styles/withStyles";
import { cohort } from "../../types";

const MARGIN_MULTIPLIER = 3;

const styles = theme => ({
  table: {
    width: "100%",
    marginTop: theme.spacing.unit * MARGIN_MULTIPLIER,
    overflowX: "auto",
    minWidth: 700
  },
  narrowCell: {
    paddingRight: theme.spacing.unit,
    paddingLeft: theme.spacing.unit,
    paddingBottom: 2,
    paddingTop: 0
  },
  button: {
    margin: theme.spacing.unit
  },
  row: {
    "&:nth-of-type(odd)": {
      backgroundColor: "#f0f0f0"
    }
  }
});

class CohortCoursesTable extends React.PureComponent {
  static propTypes = {
    classes: PropTypes.object,
    cohort: cohort,
    membersPathsRanking: PropTypes.object,
    courses: PropTypes.array,
    isInstructor: PropTypes.bool,
    onRemoveClick: PropTypes.func,
    onSortClick: PropTypes.func,
    sortState: PropTypes.shape({
      field: PropTypes.string,
      direction: PropTypes.oneOf(["asc", "desc"])
    })
  };

  render() {
    const {
      classes,
      courses: crs,
      cohort,
      isInstructor,
      onRemoveClick,
      onSortClick,
      sortState,
      membersPathsRanking,
      uid
    } = this.props;
    let totals = {
      progress: 0,
      participants: 0
    };
    const courses = crs || [];

    if (courses.length <= 0) {
      return <p>No Courses found for this cohort</p>;
    }
    courses.forEach(course => {
      totals.progress += course.progress;
      totals.participants += course.participants;
      return true;
    });
    const rankCellData = ["User Path Rank->"];
    const nonPathCellCount = 3;
    cohort.paths.forEach(path => {
      rankCellData.push(((membersPathsRanking[path] || {})[uid] || {}).rank);
    });
    for (let i = 0; i < nonPathCellCount; i++) {
      rankCellData.push("");
    }
    return (
      <Table className={classes.table}>
        <TableHead>
          <TableRow>
            <TableCell className={classes.narrowCell}>
              <TableSortLabel
                active={sortState.field === "rank"}
                direction={sortState.direction}
                onClick={() => onSortClick("rank")}
              >
                Course Rank
              </TableSortLabel>
            </TableCell>
            {cohort.pathsData && cohort.pathsData.length ? (
              cohort.pathsData.map(pathData => {
                pathData = pathData || {};
                return (
                  <TableCell
                    className={classes.narrowCell}
                    key={pathData.id || Math.random()}
                  >
                    <TableSortLabel
                      active={sortState.field === pathData.id}
                      direction={sortState.direction}
                      onClick={() => onSortClick(pathData.id)}
                    >
                      {pathData.name}
                    </TableSortLabel>
                  </TableCell>
                )
              })
            ) : (
              <TableCell className={classes.narrowCell}>
                <TableSortLabel
                  active={sortState.field === "paths"}
                  direction={sortState.direction}
                  onClick={() => onSortClick("paths")}
                >
                  Paths for Cohort
                </TableSortLabel>
              </TableCell>
            )}
            <TableCell className={classes.narrowCell}>
              <TableSortLabel
                active={sortState.field === "progress"}
                direction={sortState.direction}
                onClick={() => onSortClick("progress")}
              >
                Explorers ({totals.progress})
              </TableSortLabel>
            </TableCell>

            <TableCell className={classes.narrowCell}>
              <TableSortLabel
                active={sortState.field === "participants"}
                direction={sortState.direction}
                onClick={() => onSortClick("participants")}
              >
                Total Students ({totals.participants})
              </TableSortLabel>
            </TableCell>
            <TableCell
              className={classes.narrowCell}
              style={{ width: "50%", textAlign: "center" }}
            >
              <TableSortLabel
                active={sortState.field === "name"}
                direction={sortState.direction}
                onClick={() => onSortClick("name")}
              >
                Course
              </TableSortLabel>
            </TableCell>
            {isInstructor && <TableCell>Actions</TableCell>}
          </TableRow>
        </TableHead>
        <TableBody>
          {courses.map(course => (
              <TableRow
                className={classes.row}
                hover
                key={course.id}
                style={{ height: 18 }}
              >
                <TableCell className={classes.narrowCell}>
                  <strong>{course.rank}</strong>
                </TableCell>
                {cohort.paths ? (
                  cohort.paths.length &&
                  cohort.paths.map(id => (
                    <TableCell className={classes.narrowCell} key={id}>
                      {course.pathsProgress && course.pathsProgress[id]}
                    </TableCell>
                  ))
                ) : (
                  <TableCell className={classes.narrowCell}>None</TableCell>
                )}
                <TableCell className={classes.narrowCell}>
                  {course.progress}
                </TableCell>
                <TableCell className={classes.narrowCell}>
                  {course.participants}
                </TableCell>
                <TableCell
                  className={classes.narrowCell}
                  style={{ width: "50%", textAlign: "center" }}
                >
                  <Link to={`/courses/${course.id}`}>{course.name}</Link>
                </TableCell>
                {isInstructor && (
                  <TableCell className={classes.narrowCell}>
                    <Button
                      color="secondary"
                      onClick={() => onRemoveClick(course.id)}
                      size="small"
                    >
                      Remove
                    </Button>
                  </TableCell>
                )}
              </TableRow>
            ))}
          {membersPathsRanking && Object.keys(membersPathsRanking).length>0 &&
          <TableRow
            className={classes.row}
            hover
            style={{ height: 18, backgroundColor: "darkgrey" }}
          >
            {rankCellData.map((data, index) => {
              return (
                <TableCell
                  className={classes.narrowCell}
                  key={`${index + String(data)}`}
                >
                  {data}
                </TableCell>
              );
            })}
          </TableRow>}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(withRouter(CohortCoursesTable));
