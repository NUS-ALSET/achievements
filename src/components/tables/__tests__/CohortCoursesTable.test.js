/* eslint-disable no-magic-numbers */
import React from "react";
// import { createShallow } from "@material-ui/core/test-utils";
import sinon from "sinon";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import CohortCoursesTable from "../CohortCoursesTable";

describe("<CohortCoursesTable>", () => {
  //   let shallow;
  let mockDispatch;
  let cohort;
  let courses;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    // shallow = createShallow({
    //   dive: true
    // });
    cohort = {
      id: "test",
      name: "Test",
      description: "desc",
      paths: ["_PYBghjgnu_ghju", "_yKrfJGhnu_ghju"]
    };
    courses = [
      {
        name: "course 1",
        id: "_PofghTkjdEDyhj-"
      },
      {
        name: "course 2",
        id: "_TjJHGlkjdEDyhj-"
      }
    ];
  });

  it("renders without throwing", () => {
    const tree = renderer
      .create(
        <Router>
          <CohortCoursesTable
            cohort={cohort}
            courses={courses}
            isInstructor={true}
            membersPathsRanking={{}}
            onRemoveClick={mockDispatch}
            onSortClick={mockDispatch}
            sortState={{ field: "testfield", direction: "asc" }}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
