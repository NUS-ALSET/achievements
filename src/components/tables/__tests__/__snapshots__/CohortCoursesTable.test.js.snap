// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CohortCoursesTable> renders without throwing 1`] = `
<table
  className="MuiTable-root-5 withRouter-CohortCoursesTable--table-1"
>
  <thead
    className="MuiTableHead-root-6"
  >
    <tr
      className="MuiTableRow-root-7 MuiTableRow-head-10"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13 withRouter-CohortCoursesTable--narrowCell-2"
        scope="col"
      >
        <span
          className="MuiButtonBase-root-29 MuiTableSortLabel-root-24"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          Course Rank
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-32 MuiTableSortLabel-icon-26 MuiTableSortLabel-iconDirectionAsc-28"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13 withRouter-CohortCoursesTable--narrowCell-2"
        scope="col"
      >
        <span
          className="MuiButtonBase-root-29 MuiTableSortLabel-root-24"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          Paths for Cohort
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-32 MuiTableSortLabel-icon-26 MuiTableSortLabel-iconDirectionAsc-28"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13 withRouter-CohortCoursesTable--narrowCell-2"
        scope="col"
      >
        <span
          className="MuiButtonBase-root-29 MuiTableSortLabel-root-24"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          Explorers (
          NaN
          )
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-32 MuiTableSortLabel-icon-26 MuiTableSortLabel-iconDirectionAsc-28"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13 withRouter-CohortCoursesTable--narrowCell-2"
        scope="col"
      >
        <span
          className="MuiButtonBase-root-29 MuiTableSortLabel-root-24"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          Total Students (
          NaN
          )
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-32 MuiTableSortLabel-icon-26 MuiTableSortLabel-iconDirectionAsc-28"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13 withRouter-CohortCoursesTable--narrowCell-2"
        scope="col"
        style={
          Object {
            "textAlign": "center",
            "width": "50%",
          }
        }
      >
        <span
          className="MuiButtonBase-root-29 MuiTableSortLabel-root-24"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex="0"
        >
          Course
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-32 MuiTableSortLabel-icon-26 MuiTableSortLabel-iconDirectionAsc-28"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-head-13"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-41"
  >
    <tr
      className="MuiTableRow-root-7 MuiTableRow-hover-9 withRouter-CohortCoursesTable--row-4"
      style={
        Object {
          "height": 18,
        }
      }
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      >
        <strong />
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
        style={
          Object {
            "textAlign": "center",
            "width": "50%",
          }
        }
      >
        <a
          href="/courses/_PofghTkjdEDyhj-"
          onClick={[Function]}
        >
          course 1
        </a>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      >
        <button
          className="MuiButtonBase-root-29 MuiButton-root-42 MuiButton-text-44 MuiButton-textSecondary-46 MuiButton-flat-47 MuiButton-flatSecondary-49 MuiButton-sizeSmall-65"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-43"
          >
            Remove
          </span>
          <span
            className="MuiTouchRipple-root-68"
          />
        </button>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-7 MuiTableRow-hover-9 withRouter-CohortCoursesTable--row-4"
      style={
        Object {
          "height": 18,
        }
      }
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      >
        <strong />
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
        style={
          Object {
            "textAlign": "center",
            "width": "50%",
          }
        }
      >
        <a
          href="/courses/_TjJHGlkjdEDyhj-"
          onClick={[Function]}
        >
          course 2
        </a>
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-12 MuiTableCell-body-14 withRouter-CohortCoursesTable--narrowCell-2"
      >
        <button
          className="MuiButtonBase-root-29 MuiButton-root-42 MuiButton-text-44 MuiButton-textSecondary-46 MuiButton-flat-47 MuiButton-flatSecondary-49 MuiButton-sizeSmall-65"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-43"
          >
            Remove
          </span>
          <span
            className="MuiTouchRipple-root-68"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
